### PR TITLE
WIP: Refactor conditionals and introduced weight saving

### PR DESCRIPTION
**Pull request by alecokas**
- Moved conditional into generator
- Making use of generator inheritance
- Introduced weight saving

This is a work in progress towards introducing visualization ability, so please don't delete the branch. I think it is worth merging (after review) so that we aren't working on radically different looking class structures.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -68,6 +65,20 @@ def parse_arguments(args_to_parse):
...
+
+    evaluation = parser.add_argument_group('Evaluation settings')
+    evaluation.add_argument(
+        '--use-pretrained', action='store_true', default=False,
+        help='Load a pretrained model for inference'
+    )
+    evaluation.add_argument(
+        '--epoch-num', type=int, default=default_settings['epoch_num'],
+        help='Which checkpointed epoch number to load into memory. If set to -1, then load the most recent.'
+    )
```
Comment (Devin-Taylor) [inline main.py:81]: so many command line arguments 😂

File: main.py [hunk 2]
```diff
@@ -105,7 +116,23 @@ def main(args):
...
     # TODO: Check and see how many latent dims should be used and insert CLI argument
     if args.use_pretrained:
-        model = tf.saved_model.load(results_dir)
+        if args.epoch_num == -1:
+            # Find last checkpoint
+            args.epoch_num = extract_epoch_num(results_dir)
+
+        model = StackGAN1(
+            img_size=dataset_dims,
+            num_latent_dims=100,
```
Comment (Devin-Taylor) [inline main.py:125]: Should probably be a dynamic variable in command line/config

File: models/generators.py [hunk 3]
```diff
@@ -1,67 +1,91 @@
...
-        self.conv1 = Conv2D(filters=num_filters, kernel_size=kernel_size, activation='relu', padding='same')
-        self.upsample2 = UpSampling2D()
-        self.conv2 = Conv2D(filters=num_output_channels, kernel_size=kernel_size, activation='tanh', padding='same')
+
+        # Optimiser
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
+
+        # Weight Initialisation Parameters
+        self.w_init = tf.random_normal_initializer(stddev=0.02)
+        self.bn_init = tf.random_normal_initializer(1., 0.02)
```
Comment (Devin-Taylor) [inline models/generators.py:36]: these should be moved out of here, they are reused in the discriminator too. Either live in stackGAN or a utils file maybe, then passed as arguments?

File: models/generators.py [hunk 4]
```diff
@@ -1,67 +1,91 @@
...
+
+        # Conditional Layers
+        self.flatten = Flatten()
```
Comment (Devin-Taylor) [inline models/generators.py:39]: delete, don't think it is used

File: models/generators.py [hunk 5]
```diff
@@ -71,48 +95,59 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
...
         """
-        super().__init__()
-
-        self.w_init = tf.random_normal_initializer(stddev=0.02)
-        self.bn_init = tf.random_normal_initializer(1., 0.02)
-
-        self.img_size = img_size
-
+        super().__init__(img_size, num_latent_dims, lr, conditional_emb_size)
         num_output_channels = self.img_size[0]
```
Comment (Devin-Taylor) [inline models/generators.py:103]: you sure img_size is (3, 64, 64) and not the normal convention which is (64, 64, 3)?

**Comment by Devin-Taylor** [comment]
@alecokas looks good, small comments. Just one or two things I want you to check before merging.

File: main.py (hunk 2 above)
Comment (alecokas) [reply main.py:125]: Am I hearing more CLI arguments? :)

File: models/generators.py (hunk 3 above)
Comment (alecokas) [reply models/generators.py:36]: Ah right. Yeah I'll make that change

File: models/generators.py (hunk 5 above)
Comment (alecokas) [reply models/generators.py:103]: I'm not. It appeared that way from the existing code though. I'mm double check/

**Comment by alecokas** [comment]
I'll commit the changes this evening / afternoon

File: main.py (hunk 2 above)
Comment (alecokas) [reply main.py:125]: I'll ignore this because it actually isn't getting used yet

File: models/generators.py (hunk 3 above)
Comment (alecokas) [reply models/generators.py:36]: I'm not going to fiddle the discriminator on this PR. I'll do that right after though

File: models/generators.py (hunk 5 above)
Comment (alecokas) [reply models/generators.py:103]: Yup looks like it is so

**Comment by alecokas** [comment]
Just made the requested changes - happy for you to have another look @Devin-Taylor

**Pull request merged**